### ISSUE TITLE
Any way to ensure entire image is picked during training

### ISSUE THREAD

**Issue by sarthakpati** [fepegar/torchio#341]
Hey,

Is there any way to ensure that the full extent of an image gets picked for training? 

Cheers,
Sarthak

**Comment by fepegar** [comment]
I suppose this is related to #340. At the moment, I don't think so. I'll investigate whether this is possible with a `GridSampler`.

**Comment by romainVala** [comment]
Note that if you are using a UniformSampler, with a great number of patch, your are sure to end up picking all the images .... 
why does it not work for you ?

**Comment by sarthakpati** [comment]
> Note that if you are using a UniformSampler, with a great number of patch, your are sure to end up picking all the images ....
> why does it not work for you ?

Because I have no idea about the number of patches that would be fine to pick up the entire image (augmentations et al.).

**Comment by fepegar** [comment]
You can play with the following script to get an idea of the coverage:

```python
from itertools import islice

import torch
import torchio as tio
import numpy as np
from tqdm import tqdm
import matplotlib.pyplot as plt

# Config
patch_size = 64, 64, 64
num_patches = 300
num_runs = 5
sampler_type = 'uniform'
make_smaller = False
pad = False


if __name__ == "__main__":
    colin = tio.datasets.Colin27()
    colin.remove_image('brain')

    if pad:
        colin = tio.Pad(50)(colin)
    if make_smaller:
        patch_size = 64, 64, 5
        half = colin.spatial_shape[-1] // 2
        colin.t1['data'] = colin.t1['data'][..., half:half+5]
        colin.head['data'] = colin.head['data'][..., half:half+5]
    if sampler_type == 'uniform':
        sampler = tio.data.UniformSampler(patch_size)
    elif sampler_type == 'weighted':
        sampler = tio.data.WeightedSampler(patch_size, 'head')

    for _ in range(num_runs):
        percentages = []
        out = torch.zeros_like(colin.t1.data)
        numel = out.numel()
        for patch in tqdm(islice(sampler(colin), num_patches), total=num_patches):
            i1, j1, k1 = index_ini = np.array(patch['index_ini'])
            i2, j2, k2 = index_ini + np.array(patch_size)
            out[:, i1:i2, j1:j2, k1:k2] += 1
            ones = (out > 0).sum().item()
            percentages.append(100 * ones / numel)
        plt.plot(percentages)
    plt.grid()
    plt.title(f'Volume shape: {colin.spatial_shape}')
    plt.show()

    colin.add_image(
        tio.LabelMap(tensor=out, affine=colin.t1.affine),
        'distribution',
    )
    colin.plot()
    plt.show()
```

![Figure_2](https://user-images.githubusercontent.com/12688084/97328597-16a03780-186e-11eb-89f5-81ce041652ac.png)
![Figure_1](https://user-images.githubusercontent.com/12688084/97328613-199b2800-186e-11eb-8040-2bfc4b4a0d59.png)

Note that e.g. 300 patches could be extracted in 10 epochs using 30 `patches_per_volume` in the `Queue`.

**Comment by fepegar** [comment]
So, for now, there's no way to make sure that *all* voxels in *all* images will be sampled, but it is *quite probable* that *most* will.

**Comment by fepegar** [comment]
I suggest you play with the config parameters in the script and visualize the different results.

**Comment by romainVala** [comment]
I guess it depends on what you are training, but from my experience with segmentation learned with a patch approach, it is not a big deal to take a small number of patch (I took 8 to 16)
over the epoch you will take again and again the same data (but we different augmentation) and different patch position. 
With a random strategy you will at the end have seen almost all possible patch location so the important question is then how many epoch, which you control looking at the the loss ..

**Comment by romainVala** [comment]
@fepegar 
strange that your distribution does not look like uniform (why the center is more represented ?

**Comment by fepegar** [comment]
@romainVala intuitively, if you have an image `[x x x x x x]` and extract 4 patches with size 3 once from each index, you get:

```
[1 1 1 0 0 0]
[1 2 2 1 0 0]
[1 2 3 2 1 0]
[1 2 3 3 2 1]
```

Does that make sense?

**Comment by sarthakpati** [comment]
> So, for now, there's no way to make sure that _all_ voxels in _all_ images will be sampled, but it is _quite probable_ that _most_ will.

Well, _quite probable_ is not the same as _definitely will_. There are some architectures (for example, [DeepMedic](https://github.com/deepmedic/deepmedic)) that require **all** voxels from the input image are taken into consideration during training (yes, I know this isn't efficient and all that). This is also something that would be nice to experiment with for regression tasks...

**Comment by romainVala** [comment]
> Does that make sense?

yes I see you are plotting the number of time a specific voxel is selected, if you were ploting only the patch center, you would have the expected uniform distribution
it makes sense

**Comment by fepegar** [comment]
@sarthakpati 

I'm not sure DeepMedic needs all voxels to during training. Here's the 2015 paper: http://www.isles-challenge.org/ISLES2015/articles/kamnk1.pdf

The issue is that if you want to sample all possible locations, you have a specific `samples_per_volume` predetermined by the volume shape. I suppose you could create something like a `GridQueue` that samples *all* patches from an image.

---

@romainVala 

> if you were ploting only the patch center, you would have the expected uniform distribution

Ah, right. Let's do that, for fun:

![Figure_3](https://user-images.githubusercontent.com/12688084/97332827-cecfdf00-1872-11eb-851a-2c8f938db246.png)

**Comment by romainVala** [comment]
to come back to @sarthakpati question, if you want for sure all voxel, why do you not use the gridSampler with overlap for training ? 
https://github.com/fepegar/torchio/issues/340#issuecomment-717365204
ok it is usually used for inference, but nothing restrict it to be used for training ...

random sampler are by definition random and so to be used for a parsed sampling of the input image,  ...

**Comment by sarthakpati** [comment]
> random sampler are by definition random and so to be used for a parsed sampling of the input image

Completely agree on this

> The issue is that if you want to sample all possible locations, you have a specific samples_per_volume predetermined by the volume shape. I suppose you could create something like a GridQueue that samples all patches from an image.

Yes, this is indeed what I would be looking for...

**Comment by fepegar** [comment]
> ok it is usually used for inference, but nothing restricts it to be used for training ...

It's slightly different. The grid sampler needs the subject at initialization to compute the locations. As I said earlier, I think @sarthakpati could achieve his desired behavior using a custom implementation of the `Queue`. Feel free to submit a PR, @sarthakpati!

**Comment by sarthakpati** [comment]
I was curious if it was possible right now, but I guess not. I will try to get to it some time.
